backend-demo/src/screencast.rs

```diff
@@ -0,0 +1,113 @@
+use std::sync::{Arc, Mutex};
+
+use ashpd::{
+    backend::{
+        request::RequestImpl,
+        screencast::{
+            CreateSessionOptions, ScreencastImpl, SelectSourcesOptions, SelectSourcesResponse,
+            StartCastOptions,
+        },
+        session::CreateSessionResponse,
+        Result,
+    },
+    desktop::{
+        screencast::{CursorMode, SourceType, Stream, Streams as StartCastResponse},
+        HandleToken,
+    },
+    enumflags2::BitFlags,
+    zbus::zvariant::OwnedObjectPath,
+    AppID, PortalError, WindowIdentifierType,
+};
+use async_trait::async_trait;
+
+#[derive(Default)]
+pub struct Screencast {
+    sessions: Arc<Mutex<Vec<String>>>,
+}
+
+#[async_trait]
+impl RequestImpl for Screencast {
+    async fn close(&self, token: HandleToken) {
+        tracing::debug!("IN Close(): {token}");
+    }
+}
+
+#[async_trait]
+impl ScreencastImpl for Screencast {
+    fn available_source_types(&self) -> BitFlags<SourceType> {
+        SourceType::Monitor | SourceType::Window
+    }
+
+    fn available_cursor_mode(&self) -> BitFlags<CursorMode> {
+        CursorMode::Hidden | CursorMode::Embedded | CursorMode::Metadata
+    }
+
+    async fn create_session(
+        &self,
+        handle: OwnedObjectPath,
+        session_handle: OwnedObjectPath,
+        _app_id: Option<AppID>,
+        _options: CreateSessionOptions,
+    ) -> Result<CreateSessionResponse> {
+        tracing::debug!("IN Screencast::create_session(): {handle}, {session_handle}");
+        let session = session_handle.to_string();
+        let mut sessions = self.sessions.lock().unwrap();
+        if sessions.contains(&session) {
+            let errormsg = format!("A session with handle `{session}` already exists");
+            tracing::error!("{}", errormsg.as_str());
+            return Err(PortalError::Exist(errormsg));
+        }
+        sessions.push(session.clone());
+        Ok(CreateSessionResponse::new(session))
+    }
+
+    async fn session_closed(&self, session_handle: OwnedObjectPath) -> Result<()> {
+        tracing::debug!("IN Screencast::session_closed(): {session_handle}");
+        let session = session_handle.to_string();
+        let mut sessions = self.sessions.lock().unwrap();
+        if let Some(index) = sessions.iter().position(|x| *x == session) {
+            sessions.swap_remove(index);
+            Ok(())
+        } else {
+            let errormsg = format!("Unknown session: `{session}`");
+            tracing::error!("{}", errormsg.as_str());
+            Err(PortalError::NotFound(errormsg))
+        }
+    }
+
+    async fn select_sources(
+        &self,
+        session_handle: OwnedObjectPath,
+        _app_id: Option<AppID>,
+        _options: SelectSourcesOptions,
+    ) -> Result<SelectSourcesResponse> {
+        tracing::debug!("IN Screencast::select_sources(): {session_handle}");
+        let session = session_handle.to_string();
+        let sessions = self.sessions.lock().unwrap();
```

Comment:
I know that we don't have support for remote desktop portal yet, but see https://github.com/bilelmoussaoui/ashpd/blob/master/src/desktop/remote_desktop.rs#L33-L37. Basically, the session existence check here, should also check for session existence for RemoteDesktop sessions... Which means, somehow sharing the state between the RemoteDesktop & ScreenCast impls (probably other cases like this in the future).

I don't have a solution to this right now, but we have to find a way to store all the sessions somewhere and make it easy to retrieve a session that corresponds to a HandleToken or to close it.

Reply:
I made an attempt at addressing this future use case with a shared `SessionManager` struct that the backend builder instantiates and passes to the `ScreencastInterface` (and would pass to the `RemoteDesktopInterface` when it gets added).

Does that make sense?

Reply:
yep, seems like a decent approach till we have more constraints from other impls. Thanks!